### ISSUE TITLE
Discrepancy in Generation Metrics: Results Differ from Table 1 in Your Paper

### ISSUE THREAD

**Issue by HuangWanqiu** [dongyangli-del/EEG_Image_decode#39]
Hello, I would like to sincerely thank you for your excellent work and for sharing the detailed code.

I used the generated images shared in your Google Drive (https://drive.google.com/drive/folders/1SAdAs73-w_N9ZVy2xAMsJZyyIE28Fnes?usp=sharing) to evaluate the generation metrics using the `Generation_metrics_sub8.ipynb` notebook. However, **the results I obtained differ from those reported in Table 1 of your paper**. 
**Could you please help identify any mistakes in my evaluation process**? 
Thank you!

### Modifications
The only modification I made in `Generation_metrics_sub8.ipynb` was the way I generated the `all_images.pt` file. Since the `generated images` folder in Google Drive contains 10 images per class, I randomly selected one image from each class to generate `all_images.pt`. Below is the specific code I used:

```python
# Initialize a list to hold all the image tensors
tensor_list = []

# Iterate over the folders in the source directory
for folder_name in os.listdir(source_dir):
    folder_path = os.path.join(source_dir, folder_name)
    
    # Check if it's a directory
    if os.path.isdir(folder_path):
        # Iterate over the images in the folder
        for image_name in os.listdir(folder_path):
            image_path = os.path.join(source_dir, folder_name, image_name)
            
            # Load the image
            with Image.open(image_path) as img:
                # Convert the image to a PyTorch tensor and add a batch dimension
                tensor = torch.tensor(np.array(img)).unsqueeze(0)
                tensor_list.append(tensor)
            if len(tensor_list) >= 1: break # NOTE: for test

# Concatenate all tensors along the 0th dimension
all_tensors = torch.cat(tensor_list, dim=0)

# Save the combined tensor
combined_tensor_path = os.path.join(target_dir, "all_images.pt")
torch.save(all_tensors, combined_tensor_path)
```

### Issue
Even with this minor modification, the evaluation results do not match the reported metrics in your paper. Below are the results I obtained:

```
Distance: 0.6946313013136387
     Metric    Value
    PixCorr 0.097653
       SSIM 0.308444
 AlexNet(2) 0.503618
 AlexNet(5) 0.548643
InceptionV3 0.496080
       CLIP 0.522337
   EffNet-B 0.979708
       SwAV 0.694631
```

![Image](https://github.com/user-attachments/assets/408a5f07-b011-4743-a259-2a2e8f22a696)

**Comment by x2x5** [comment]
下面的第二列是我用作者提供的sub08生成的图片跑的结果，第三列我是自己用作者的代码训练再生成图片跑的结果。

第二列还是挺接近的。

| 指标 (Metric) | 论文结果 | 复现结果 (使用作者图片) | 自训练模型测试结果 |
| :--- | :--- | :--- | :--- |
| **低层级指标 (Low-level)** | | | |
| PixCorr | 0.160 | 0.162 | 0.163 |
| SSIM | 0.345 | 0.371 | **0.422** |
| **高层级/语义指标 (High-level/Semantic)** | | | |
| AlexNet(2) | 0.776 | **0.780** | 0.607 |
| AlexNet(5) | 0.866 | **0.866** | 0.615 |
| Inception / InceptionV3 | 0.734 | **0.742** | 0.543 |
| CLIP | 0.786 | **0.789** | 0.567 |
| SwAV | 0.582 | 0.580 | **0.654** |
| EffNet-B | - | 0.880 | **0.943** |

**Comment by AIyumeng** [comment]
I obtained the following metrics:

```
     Metric    Value
    PixCorr 0.156171
       SSIM 0.344204
 AlexNet(2) 0.806156
 AlexNet(5) 0.865729
InceptionV3 0.739824
       CLIP 0.775779
   EffNet-B 0.880971
       SwAV 0.586844
```
I think the issue comes from the code snippet below:
```
# Initialize a list to hold all the image tensors
tensor_list = []
dir_list = os.listdir(source_dir)
dir_list.sort()
# Iterate over the folders in the source directory
for folder_name in dir_list:
    ...
```

The problem is that os.listdir() does not guarantee a consistent order across runs, which can cause the metric mismatch.
Sorting the directory list before iterating (using dir_list.sort()) should fix this issue.

**Comment by DuDai** [comment]
> I obtained the following metrics:
> 
> ```
>      Metric    Value
>     PixCorr 0.156171
>        SSIM 0.344204
>  AlexNet(2) 0.806156
>  AlexNet(5) 0.865729
> InceptionV3 0.739824
>        CLIP 0.775779
>    EffNet-B 0.880971
>        SwAV 0.586844
> ```
> 
> I think the issue comes from the code snippet below:
> 
> ```
> # Initialize a list to hold all the image tensors
> tensor_list = []
> dir_list = os.listdir(source_dir)
> dir_list.sort()
> # Iterate over the folders in the source directory
> for folder_name in dir_list:
>     ...
> ```
> 
> The problem is that os.listdir() does not guarantee a consistent order across runs, which can cause the metric mismatch. Sorting the directory list before iterating (using dir_list.sort()) should fix this issue.

Can you share your code? I really need it, thanks

**Comment by DuDai** [comment]
> https://drive.google.com/file/d/164IbMDYXPsH0nyNGgjdkBA7SxK1Hglhx/view?usp=sharing
> 
> sorry, now it's ok.

The code issue has been resolved. Thank you very much!
